cpp/include/experimental/graph.hpp

```diff
@@ -0,0 +1,198 @@
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+#pragma once
+
+#include <experimental/graph_view.hpp>
+#include <utilities/error.hpp>
+
+#include <raft/handle.hpp>
+#include <rmm/device_uvector.hpp>
+
+#include <cstddef>
+#include <string>
+#include <type_traits>
+#include <vector>
+
+namespace cugraph {
+namespace experimental {
+
+template <typename vertex_t, typename edge_t, typename weight_t>
+struct edgelist_t {
+  vertex_t const *p_src_vertices{nullptr};
+  vertex_t const *p_dst_vertices{nullptr};
+  weight_t const *p_edge_weights{nullptr};
+  edge_t number_of_edges{0};
+};
+
+// graph_t is an owning graph class (note that graph_view_t is a non-owning graph class)
+template <typename vertex_t,
+          typename edge_t,
+          typename weight_t,
+          bool store_transposed,
+          bool multi_gpu,
+          typename Enable = void>
+class graph_t;
+
+// multi-GPU version
+template <typename vertex_t,
+          typename edge_t,
+          typename weight_t,
+          bool store_transposed,
+          bool multi_gpu>
+class graph_t<vertex_t, edge_t, weight_t, store_transposed, multi_gpu, std::enable_if_t<multi_gpu>>
+  : public detail::graph_base_t<vertex_t, edge_t, weight_t> {
+ public:
+  using vertex_type                              = vertex_t;
+  using edge_type                                = edge_t;
+  using weight_type                              = weight_t;
+  static constexpr bool is_adj_matrix_transposed = store_transposed;
+  static constexpr bool is_multi_gpu             = multi_gpu;
+
+  graph_t(raft::handle_t const &handle,
+          std::vector<edgelist_t<vertex_t, edge_t, weight_t>> const &edge_lists,
+          partition_t<vertex_t> const &partition,
+          vertex_t number_of_vertices,
+          edge_t number_of_edges,
+          bool is_symmetric,
+          bool is_multigraph,
+          bool is_weighted,
+          bool sorted_by_global_degree_within_vertex_partition,
+          bool do_expensive_check = false);
+
+  vertex_t get_number_of_local_vertices() const
+  {
+    return partition_.get_vertex_partition_range_last() -
+           partition_.get_vertex_partition_range_first();
+  }
+
+  graph_view_t<vertex_t, edge_t, weight_t, store_transposed, multi_gpu> view()
+  {
+    std::vector<edge_t const *> offsets(adj_matrix_partition_offsets_.size(), nullptr);
+    std::vector<vertex_t const *> indices(adj_matrix_partition_indices_.size(), nullptr);
+    std::vector<weight_t const *> weights(adj_matrix_partition_weights_.size(), nullptr);
+    for (size_t i = 0; i < offsets.size(); ++i) {
+      offsets[i] = adj_matrix_partition_offsets_[i].data();
+      indices[i] = adj_matrix_partition_indices_[i].data();
+      if (this->is_weighted()) { weights[i] = adj_matrix_partition_weights_[i].data(); }
```

Comment:
If weights are passed but is_weighted is false, then, exception is raised (this is part of input argument check).

The question is whether we want to handle is_weighted same to other graph properties (e.g. is_symmetric, is_multigraph) or want to handle this differently and instead of keeping this value, we can compute this when asked (thankfully, checking weights is nullptr is cheap, but checking a graph has multi-edges or symmetric is not).

What do you think?

Reply:
I think the graph class responsibility is just to store what's given and infer if the graph is weighted or not. Later, algos may or may not try to use weights and exceptions should be triggered at that time.

Reply:
Done.